components/phenotype-suggest-service/api/src/main/java/org/phenotips/solr/PhenotypeSuggestService.java

```diff
@@ -148,20 +151,56 @@ protected String getName()
      * @param nphenotypes phenotypes that are not observed in the patient
      * @return the computed Solr query parameters
      */
-    private MapSolrParams prepareParams(Collection<String> phenotypes, Collection<String> nphenotypes)
+    private SolrQuery prepareParams(Collection<String> phenotypes, Collection<String> nphenotypes)
     {
-        Map<String, String> params = new HashMap<>();
+        SolrQuery result = new SolrQuery();
         String q = "symptom:" + StringUtils.join(phenotypes, " symptom:");
         if (!nphenotypes.isEmpty()) {
             q += "  not_symptom:" + StringUtils.join(nphenotypes, " not_symptom:");
         }
         q += " -nameSort:\\** -nameSort:\\+* -nameSort:\\^*";
-        params.put(CommonParams.Q, q.replaceAll("HP:", "HP\\\\:"));
-        params.put(CommonParams.ROWS, "100");
-        params.put(CommonParams.START, "0");
-        params.put(CommonParams.DEBUG_QUERY, Boolean.toString(true));
-        params.put(CommonParams.EXPLAIN_STRUCT, Boolean.toString(true));
+        result.set(CommonParams.Q, q.replaceAll("HP:", "HP\\\\:"));
+        result.set(CommonParams.ROWS, "100");
+        result.set(CommonParams.START, "0");
+        result.set(CommonParams.DEBUG_QUERY, Boolean.toString(true));
+        result.set(CommonParams.EXPLAIN_STRUCT, Boolean.toString(true));
+
+        return result;
+    }
 
-        return new MapSolrParams(params);
+    /**
+     * Get the HPO IDs of the specified phenotype and all its ancestors.
+     *
+     * @param id the HPO identifier to search for, in the {@code HP:1234567} format
+     * @return the full set of ancestors-or-self IDs, or an empty set if the requested ID was not found in the index
+     */
+    @SuppressWarnings("unchecked")
+    public Set<String> getAllAncestorsAndSelfIDs(final String id)
```

Comment:
A vocabulary term already has a `term.getAllAncestorsAndSelf` method. Instead of duplicating code, you should just call that method, then collect the IDs of each ancestor term.